internal/controller/standalone_pgadmin/controller.go

```diff
@@ -59,3 +116,343 @@ func (r *PGAdminReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	cannotResize := func(err error) {
+		meta.SetStatusCondition(&pgadmin.Status.Conditions, metav1.Condition{
+			Type:    v1beta1.PersistentVolumeResizing,
+			Status:  metav1.ConditionFalse,
+			Reason:  string(apierrors.ReasonForError(err)),
+			Message: "One or more volumes cannot be resized",
+
+			ObservedGeneration: pgadmin.Generation,
+		})
+	}
+
+	volumeError := func(err error) {
+		r.Recorder.Event(pgadmin,
+			corev1.EventTypeWarning, "PersistentVolumeError", err.Error())
+	}
+
+	// Forbidden means (RBAC is broken or) the API request was rejected by an
+	// admission controller. Assume it is the latter and raise the issue as a
+	// condition and event.
+	// - https://releases.k8s.io/v1.21.0/plugin/pkg/admission/storage/persistentvolume/resize/admission.go
+	if apierrors.IsForbidden(err) {
+		cannotResize(err)
+		volumeError(err)
+		return nil
+	}
+
+	if apierrors.IsInvalid(err) && status.Details != nil {
+		unknownCause := false
+		for _, cause := range status.Details.Causes {
+			switch {
+			// Forbidden "spec" happens when the PVC is waiting to be bound.
+			// It should resolve on its own and trigger another reconcile. Raise
+			// the issue as an event.
+			// - https://releases.k8s.io/v1.21.0/pkg/apis/core/validation/validation.go#L2028
+			//
+			// TODO(cbandy): This can also happen when changing a field other
+			// than requests within the spec (access modes, storage class, etc).
+			// That case needs a condition or should be prevented via a webhook.
+			case
+				cause.Type == metav1.CauseType(field.ErrorTypeForbidden) &&
+					cause.Field == "spec":
+				volumeError(err)
+
+			// Forbidden "storage" happens when the change is not allowed. Raise
+			// the issue as a condition and event.
+			// - https://releases.k8s.io/v1.21.0/pkg/apis/core/validation/validation.go#L2028
+			case
+				cause.Type == metav1.CauseType(field.ErrorTypeForbidden) &&
+					cause.Field == "spec.resources.requests.storage":
+				cannotResize(err)
+				volumeError(err)
+
+			default:
+				unknownCause = true
+			}
+		}
+
+		if len(status.Details.Causes) > 0 && !unknownCause {
+			// All the causes were identified and handled.
+			return nil
+		}
+	}
+
+	return err
+}
+
+// The owner reference created by controllerutil.SetControllerReference blocks
+// deletion. The OwnerReferencesPermissionEnforcement plugin requires that the
+// creator of such a reference have either "delete" permission on the owner or
+// "update" permission on the owner's "finalizers" subresource.
+// - https://docs.k8s.io/reference/access-authn-authz/admission-controllers/
+// +kubebuilder:rbac:groups="postgres-operator.crunchydata.com",resources="postgresclusters/finalizers",verbs={update}
+
+// setControllerReference sets owner as a Controller OwnerReference on controlled.
+// Only one OwnerReference can be a controller, so it returns an error if another
+// is already set.
+//
+// TODO(tjmoore4): This function is duplicated from a version that takes a PostgresCluster object.
```

Comment:
is there a way to make the functions more generic :thinking:

Reply:
Possibly in the future. Per discussion, leave as is for now.